### ISSUE TITLE
Rule Change: support `Object.groupBy` and `Map.groupBy` in `array-callback-return`

### ISSUE THREAD

**Issue by synthex-byte** [eslint/eslint#21113]
### What rule do you want to change?

array-callback-return

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/* eslint array-callback-return: "error" */
const grouped = Object.groupBy(inventory, ({ type }) => {
    if (type === "fruit") {
        return "fruit";
    }
    // forgot to return
});
```

### What does the rule currently do for this code?

No problem is currently reported. At runtime, elements whose callback returns undefined are silently grouped under the key `"undefined"`

### What will the rule do after it's changed?

Report a problem when the callback passed to `Object.groupBy` or `Map.groupBy` doesn't return a value on some code path, consistent with the existing behavior for `Array.from` . The callback is the second argument, same as Array.from / Array.fromAsync

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

[Object.groupBy](https://tc39.es/ecma262/#sec-object.groupby) and [Map.groupBy](https://tc39.es/ecma262/#sec-map.groupby) have reached TC39 stage 4 and are part of ES2024, so I think it makes sense for this rule to cover them as well. Similar to what was done for Array.fromAsync (#20438)

**Comment by aladdin-add** [comment]
it makes sense! please feel free to send a PR. 👍

**Comment by xevemuy** [comment]
@aladdin-add I don't think arrays have a `groupBy` method. Could you check again?

**Comment by aladdin-add** [comment]
oh, sorry I missed it. Although the example makes sense, it does not seem to fit this rule name `array-callback-return`. thoughts?

**Comment by synthex-byte** [comment]
I think the rule already covers static methods like Array.from and Array.fromAsync, so although groupBy isn’t an Array method, this change seems similar in purpose: checking callbacks passed to built-in methods. Is there any concern with handling Object.groupBy and Map.groupBy in this rule?

**Comment by synthex-byte** [comment]
Friendly ping for consensus on this suggestion.

**Comment by aladdin-add** [comment]
ping @eslint/eslint-team

**Comment by mdjermanovic** [comment]
Since this rule’s name and documentation explicitly state that it applies to array methods, including non-array methods would be unexpected.

I think we could consider adding a new rule or renaming the existing one to a name that doesn't include "array", which could then be extended to support these methods.

**Comment by synthex-byte** [comment]
The naming concern makes sense. Which method would the team prefer here — a new rule, or renaming the existing one? Also, are we still waiting for more feedback on this? Just asking since it's still on the feedback board 😊

**Comment by synthex-byte** [comment]
ping ?

**Comment by synthex-byte** [comment]
With the next TSC meeting coming up, could we add this to the agenda to decide on the direction (creating a new rule vs renaming/deprecating array-callback-return)?

**Comment by synthex-byte** [comment]
Hi its been quite some time as previously discussed in the tsc metting the team reached the conclusion of moving with the new rule that include these method also as @mdjermanovic said that the name can be discussed on the issue so i was hoping to get it moving

**Comment by eslintbot** [comment]
Resolution from the ESLint TSC meeting on 2026-09-03: The TSC agreed that `Object.groupBy` and `Map.groupBy` do not belong in `array-callback-return`, and that a new, generic rule should be created to cover both array callbacks and these new methods. Further discussion on the new rule's name and exact scope will continue on the issue.

> Generated by [ESLint TSC Meeting Notes & Follow-up](https://github.com/eslint/tsc-meetings/actions/runs/34648564892) for #749 · copilot · auto · 29.6 AIC · ⌖ 6.3 AIC · ⊞ 7.4K · [◷](https://github.com/search?q=repo%3Aeslint%2Feslint+%22gh-aw-workflow-call-id%3A+eslint%2Ftsc-meetings%2Fgenerate-meeting-notes%22&type=issues)
